ql/src/java/org/apache/hadoop/hive/ql/parse/TezCompiler.java

```diff
@@ -1046,11 +1046,16 @@ public Object process(Node nd, Stack<Node> stack, NodeProcessorCtx procCtx, Obje
         }
       }
       if (nd instanceof TableScanOperator) {
+        TableScanOperator ts = (TableScanOperator) nd;
         // If the tablescan operator is making use of filtering capabilities of readers then
         // we will not see the actual incoming rowcount which was processed - so we may not use it for relNodes
-        TableScanOperator ts = (TableScanOperator) nd;
         if (ts.getConf().getPredicateString() != null) {
-          planMapper.link(ts, new OperatorStats.MayNotUseForRelNodes());
+          invalidateForRelNodes(ts, false);
+        }
+        // If sampling is configured, the table scan could be canceled in the middle. We avoid using runtime stats
+        // for HiveTableScan and its descendants as it is not pushed down to HiveTableScan RelNodes
+        if (ts.getConf().getRowLimit() >= 0) {
+          invalidateForRelNodes(ts, true);
```

Comment:
This might be too defensive as currently no RelNodes are created when TABLESAMPLE is used.

Reply:
@okumin, is that possible/OK to link multiple MayNotUseForRelNodes to the same TableScanOperator

Reply:
👍